emote/models/model.py

```diff
@@ -0,0 +1,241 @@
+                self.model.device,
+                dtype=torch.double if normalize_double_precision else torch.float,
+            )
+        self.device = self.model.device
+        self.learned_rewards = learned_rewards
+        self.target_is_delta = target_is_delta
+        self.no_delta_list = no_delta_list if no_delta_list else []
+        self.obs_process_fn = obs_process_fn
+
+    def forward(self, x: torch.Tensor, *args, **kwargs) -> tuple[torch.Tensor, ...]:
+        """Computes the output of the dynamics model.
+        Args:
+            x (tensor): input
+        Returns:
+            (tuple of tensors): predicted tensors
+        """
+        return self.model.forward(x, *args, **kwargs)
+
+    def loss(
+        self,
+        obs: torch.Tensor,
+        next_obs: torch.Tensor,
+        action: torch.Tensor,
+        reward: torch.Tensor,
+    ) -> tuple[torch.Tensor, dict[str, any]]:
+        """Computes the model loss over a batch of transitions.
+        Args:
+            obs (tensor): current observations
+            next_obs (tensor): next observations
+            action (tensor): actions
+            reward (tensor): rewards
+        Returns:
+            (tensor and optional dict): as returned by `model.loss().`
+        """
+        model_in, target = self.process_batch(
+            obs=obs, next_obs=next_obs, action=action, reward=reward
+        )
+        return self.model.loss(model_in, target=target)
+
+    def sample(
+        self,
+        action: torch.Tensor,
+        observation: torch.Tensor,
+        rng: torch.Generator,
+    ) -> tuple[torch.Tensor, Optional[torch.Tensor]]:
+        """Samples a simulated transition from the dynamics model.
+        Args:
+            action (tensor): the action at.
+            observation (tensor): the observation/state st.
+            rng (torch.Generator): a random number generator.
+        Returns:
+            (tuple): predicted observation and rewards.
+        """
+        obs = to_tensor(observation).to(self.device)
+        model_in = self.get_model_input(obs, action)
+        if not hasattr(self.model, "sample"):
+            raise RuntimeError(
+                "DynamicModel requires wrapped model to implement sample method"
+            )
+        preds = self.model.sample(model_in, rng)
+        if len(preds.shape) != 2:
+            raise RuntimeError(
+                "Prediction shape is: {} "
+                "Predictions must be 'batch_size x length_of_prediction'."
+                "Have you forgotten to run propagation on the ensemble?".format(
+                    preds.shape
+                )
+            )
+        next_observs = preds[:, :-1] if self.learned_rewards else preds
+        if self.target_is_delta:
+            tmp_ = next_observs + obs
+            for dim in self.no_delta_list:
+                tmp_[:, dim] = next_observs[:, dim]
+            next_observs = tmp_
+        rewards = preds[:, -1:] if self.learned_rewards else None
```

Comment:
The main idea of having a wrapper class is for the model not to care about state, actions and rewards. The model simply is a regression model that maps some inputs to some outputs :) So delegating knowing reward to model is not inline with this approach. 
To simplify the code, I removed the "target_is_delta" flag, and assume that target is always delta. Then, we'll only have no_delta_mask for the dimensions of the obs that should not be delta.

Reply:
masking is a bit more complicated than a list of dimension because of the batch size.

Reply:
Alrighty, makes sense! I figured with broadcasting it might work out fine (maybe bmm...) but if you think this is better/easier/clearer I'll defer to you :-)

Reply:
I resolve it for now to be able to continue. Please feel free to bring it up again if needed